rust/src/shards/gui/containers/docking.rs

```diff
@@ -0,0 +1,441 @@
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::types::Context;
+use crate::types::ExposedInfo;
+use crate::types::ExposedTypes;
+use crate::types::InstanceData;
+use crate::types::OptionalString;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::Seq;
+use crate::types::ShardRef;
+use crate::types::ShardsVar;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::ANY_TYPES;
+use crate::types::SEQ_OF_SHARDS_TYPES;
+use crate::types::SHARDS_OR_NONE_TYPES;
+use crate::types::STRING_TYPES;
+
+const TABS_NAME: &'static str = "UI.Tabs";
+
+lazy_static! {
+  static ref TAB_PARAMETERS: Parameters = vec![
+    (
+      cstr!("Title"),
+      cstr!("The title of the tab."),
+      &STRING_TYPES[..],
+    )
+      .into(),
+    (
+      cstr!("Contents"),
+      cstr!("The UI contents."),
+      &SHARDS_OR_NONE_TYPES[..],
+    )
+      .into(),
+  ];
+  static ref DOCKAREA_PARAMETERS: Parameters = vec![
+    (
+      cstr!("Contents"),
+      cstr!("The UI contents for each tab."),
+      &SEQ_OF_SHARDS_TYPES[..],
+    )
+      .into(),
+    (
+      cstr!("Titles"),
+      cstr!("The titles for each tab."),
+      &SEQ_OF_SHARDS_TYPES[..],
+    )
+      .into(),
+  ];
+}
+
+impl Default for Tab {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      title: ParamVar::default(),
+      contents: ShardsVar::default(),
+      exposing: Vec::new(),
+    }
+  }
+}
+
+impl Shard for Tab {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Tab")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Tab-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    TABS_NAME
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Represents a tab inside a DockArea."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!(
+      "The value that will be passed to the Contents shards of the tab."
+    ))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("The output of this shard will be its input."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&TAB_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => Ok(self.title.set_param(value)),
+      1 => self.contents.set_param(value),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.title.get_param(),
+      1 => self.contents.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
```

Comment:
Shouldn't this also expose? or it can be "hidden" / "not execute" conditionally?

Reply:
Yeah. Likely forgot expose here. Since we can't close tabs for now, they are not executed conditionally.